articles/design-patterns/index.md

```diff
@@ -0,0 +1,213 @@
+# Understanding design patterns in Java
+
+### What are design patterns?
+
+A design pattern is a reusable solution to a commonly occurring problem in software design.
+They can be thought of as tried and tested approaches; guidelines for how to structure your objects and the relationships between them. These guidelines are not too opinionated, they are implementation specific and are meant to be used with any programming language while designing any kind of system.
+Design patterns were invented by a group of software engineers known as The Gang of Four in their book, Elements of Reusable Object-Oriented Software. They created 23 design patterns and revolutionized the way people design software. 
+Though the book was written in a C++ context, it is still relevant to Java programmers.
+Let us discuss some of the most popular design patterns and how to implement them in Java using simple examples.
+
+### The Singleton pattern
+The Singleton pattern is perhaps the most popular and controversial design pattern. 
+It is the design pattern where only one instance of a class is allowed to exist globally. This single instance is used wherever it is needed without creating a new one.
+The class is responsible for preventing any other objects from creating new instances of it. It is also responsible for presenting the single instance to any class that might need it.
+This pattern is useful in situations like when working with a database where having more than one object might bring about inconsistencies.
+Let's see how we can implement it in Java:
+```java
+public class Database {
+    private static Database _instance = null;
+
+    private Database() { //make the constructor private so that only code that's inside this class can call it.
+    }
+
+    public static Database getInstance() {
+        if (_instance == null) { // check if the instance already exists. If it has not been instantiated, create the instance.
+            _instance = new Database();
+        }
+
+        return _instance; //if the instance is not empty, return the instance that exists.
+    }
+}
+```
+
+When you are using multiple threads, this style of creating a Singleton is not the best; multiple threads might get a hold of different instances of our singleton.
+There are two ways we can avoid this:
+1.  Find a way to make the getInstance method synchronized.
+2.  Create the instance of the Singleton class as soon as the class is loaded to the JVM.
+#### Limitations of the Singleton pattern
+Singletons introduce a global state. This brings about tight coupling that makes unit testing difficult.
```

Comment:
You may need to briefly explain what "coupling" is. If this is meant to be read by people who don't know much about design patterns, then they probably don't know what coupling is.

Reply:
@botahamec  done.